spec/support/shared/examples/msf/core/exploit/sqli/sqli_common_spec.rb

```diff
@@ -0,0 +1,155 @@
+RSpec.shared_examples 'Msf::Exploit::SQLi::Common' do |sqli_class|
+  let(:common_class) do
+    sqli_class
+  end
+  before(:example) do
+    # because vprint_status accesses framework, datastore and user_output
+    allow_any_instance_of(common_class).to receive(:vprint_status).and_return(nil)
+  end
+
+  context 'Without opts' do
+    let(:sqli_obj) do
+      common_class.new({}, {}, {}) do |payload|
+        payload[/'(.+?)'/, 1] || ''
+      end
+    end
+    context('#run_sql') do
+      queries = ["select concat(username,':',password) from users", "select 'hello'", 'select 1234 from users']
+      query_results = [ ':', 'hello', '' ]
+      queries.each_with_index do |query, i|
+        it 'Should call vprint_status on run_sql' do
+          expect(sqli_obj).to receive(:vprint_status).once
+          expect(sqli_obj.run_sql(query)).to eql query_results[i]
+        end
+      end
+    end
+
+    context('#test_vulnerable') do
+      it 'Should detect if the sqli object is expected to perform SQLi successfully' do
+        expect(sqli_obj.test_vulnerable).to eql true
+        allow(sqli_obj).to receive(:run_sql).and_return '<div id="articles"></div>'
+        expect(sqli_obj.test_vulnerable).to eql false
+      end
+    end
+
+    context('#dump_table_fields') do
+      let(:query_proc) do
+        sqli_obj.instance_variable_get(:@query_proc)
+      end
+      result_limit = rand(1..26)
+      common_query = /^select.*password.*from.*maindb\.users\s*;?\s*?(?:#|--)?$/mi
+      condition_query = /^select.*password.*from.*maindb\.users\s+where/mi
+      limit_query = /^select.*password.*from.*maindb\.users\s+limit/mi
+
+      # query without condition and limit
+      it 'Should yield valid queries' do
+        expect(sqli_obj).to receive(:run_sql).and_call_original
+        expect(query_proc).to receive(:call).with(common_query).and_call_original
+        sqli_obj.dump_table_fields('maindb.users', %w[password])
+      end
+      # query with condition string
+      it 'Should yield valid queries when the user adds a condition' do
+        expect(query_proc).to receive(:call).with(condition_query).and_call_original
+        sqli_obj.dump_table_fields('maindb.users', %w[password], "username='admin'")
+      end
+      # query with limit
+      it 'Should yield valid queries when the user adds a limit number' do
+        expect(query_proc).to receive(:call).with(limit_query).and_call_original
+        sqli_obj.dump_table_fields('maindb.users', %w[password], '', result_limit)
```

Comment:
I'm not familiar with the internals of `sqli_obj.dump_table_fields`- is there any output that could be verified as part of this test?

Reply:
`dump_table_fields` should return an Array of rows (each row being an array of strings), the dump of a table from the database.

It just generates some SQL queries (depending on the user options) and calls `run_sql` (or truncated_query), which yield the SQL to the user-supplied block (`query_proc`), which is expected to query the server and return the response, for this reason, it's hard to check its results without having a database set-up (and validating the DBMS-specific SQL also is hard, for this, I'm using some regular expressions that just check for basic things that should be there), but the return value really depends on what the server returned (for `select group_concat(table_name) from information_schema.tables where table_schema=database()` for example, how would I check what the expected return value would be?)